pkg/search/controller.go

```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
+		for resource, previousRegistries := range cr.resources {
+			delete(previousRegistries, name)
+			if len(previousRegistries) == 0 {
+				removedResources = append(removedResources, resource)
+			}
+		}
+		for _, resource := range removedResources {
+			delete(cr.resources, resource)
+		}
+	}
+	for _, registry := range matched {
+		addedResources = c.getRegistryAddedResources(registry, cr, addedResources)
+	}
+	return
+}
+
+func (c *Controller) findOpenSearchBackend(matchedRegistries map[string]*searchv1alpha1.ResourceRegistry) *searchv1alpha1.BackendStoreConfig {
+	for _, registry := range matchedRegistries {
+		if backend := registry.Spec.BackendStore; backend != nil && backend.OpenSearch != nil {
+			// one cluster may be related to multi registries,
+			// however only one backend could keep in memory with one cluster
+			return backend
+		}
 	}
+	return nil
+}
 
-	// STEP1: stop informer manager for the cluster which does not exist anymore or is not ready.
-	cls, err := c.clusterLister.Get(cluster)
+func (c *Controller) clusterAbleToCache(cluster string) (cls *clusterv1alpha1.Cluster, able bool, err error) {
+	cls, err = c.clusterLister.Get(cluster)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			klog.Infof("try to stop cluster informer %s", cluster)
+			klog.Infof("Try to stop cluster informer %s", cluster)
 			c.InformerManager.Stop(cluster)
-			return nil
+			return
 		}
-		return err
+		return
 	}
 
 	if !cls.DeletionTimestamp.IsZero() {
-		klog.Infof("try to stop cluster informer %s", cluster)
+		klog.Infof("Try to stop cluster informer %s", cluster)
 		c.InformerManager.Stop(cluster)
-		return nil
+		return
 	}
 
 	if !util.IsClusterReady(&cls.Status) {
 		klog.Warningf("cluster %s is notReady try to stop this cluster informer", cluster)
 		c.InformerManager.Stop(cluster)
+		return
+	}
+	able = true
+	return
+}
+
+func (c *Controller) reconcileClusterWithRegistries(cls *clusterv1alpha1.Cluster) (matchedRegistries map[string]*searchv1alpha1.ResourceRegistry, cr *clusterRegistry, resourcesChanged, newRegistry bool, err error) {
+	cluster := cls.Name
+	var allRegistries map[string]*searchv1alpha1.ResourceRegistry
+	allRegistries, matchedRegistries, err = c.getClusterMatchedRegistries(cls)
+	if err != nil {
+		klog.Errorf("Get cluster matched resource registries failed, cluster: %s, error: %s", cluster, err)
+		return
+	}
+	klog.V(4).Infof("Registries matched with cluster, cluster: %s, registries: %s",
+		cluster, strings.Join(util.Keys(matchedRegistries), ", "))
+	v, hasExistingRegistries := c.clusterRegistry.Load(cluster)
+	var addedRegistries, removedRegistries []string
+	if !hasExistingRegistries {
+		klog.Infof("Cluster %s has no existing registries", cluster)
+		addedRegistries, removedRegistries = util.DiffKey(registrySet(nil), matchedRegistries)
+	} else {
+		crv := v.(clusterRegistry)
+		cr = &crv
+		addedRegistries, removedRegistries = util.DiffKey(cr.registries, matchedRegistries)
+	}
+	if len(addedRegistries) > 0 {
```

Comment:
Can we omit these judgments to improve the code readability and reduce cyclomatic complexity? Same as below.

Reply:
Yes, at first I thought it's not need to stop the informer, just remove uneeded resources. So I did diff to find out what are added, what are removed.

But it's not work, we have to stop a informer if something changed, to init a new one. Unless we do a huge refactor of informer.